cpp/src/quantiles/quantiles.cu

```diff
@@ -0,0 +1,279 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "cudf/legacy/reduction.hpp"
+#include "cudf/types.hpp"
+#include "cudf/utilities/bit.hpp"
+#include "thrust/iterator/transform_iterator.h"
+#include "thrust/transform.h"
+#include <cudf/copying.hpp>
+#include <cudf/sorting.hpp>
+#include <cudf/utilities/error.hpp>
+#include <iostream>
+#include <quantiles/legacy/quantiles_util.hpp>
+#include <rmm/thrust_rmm_allocator.h>
+#include <thrust/extrema.h>
+#include <thrust/sort.h>
+#include <algorithm>
+#include <cmath>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/scalar/scalar.hpp>
+#include <cudf/scalar/scalar_factories.hpp>
+#include <memory>
+#include <stdexcept>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/copying.hpp>
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+namespace {
+
+template <typename T>
+CUDA_HOST_DEVICE_CALLABLE
+T get_array_value(T const* devarr, size_type location)
+{
+    T result;
+#if defined(__CUDA_ARCH__)
+    result = devarr[location];
+#else
+    CUDA_TRY( cudaMemcpy(&result, devarr + location, sizeof(T), cudaMemcpyDeviceToHost) );
+#endif
+    return result;
+}
+
+struct quantile_index
+{
+    size_type lower_bound;
+    size_type upper_bound;
+    size_type nearest;
+    double fraction;
+
+    quantile_index(size_type count, double quantile)
+    {
+        quantile = std::min(std::max(quantile, 0.0), 1.0);
+
+        double val = quantile * (count - 1);
+        lower_bound = std::floor(val);
+        upper_bound = static_cast<size_t>(std::ceil(val));
+        nearest = static_cast<size_t>(std::nearbyint(val));
+        fraction = val - lower_bound;
+    }
+};
+
+} // anonymous namespace
+
+template<typename T, typename TResult, typename TSortMap, bool sortmap_is_gpu>
+std::unique_ptr<scalar>
+select_quantile(T const * begin,
+                TSortMap sortmap,
+                size_t size,
+                double quantile,
+                interpolation interpolation)
+{
+    
+    if (size < 2) {
+        auto result_value = get_array_value(begin, 0);
+        return std::make_unique<numeric_scalar<TResult>>(result_value);
+    }
+
+    quantile_index idx(size, quantile);
+
+    T a;
+    T b;
+    TResult value;
+
+    switch (interpolation) {
+    case interpolation::LINEAR:
+        a = get_array_value(begin, sortmap[idx.lower_bound]);
+        b = get_array_value(begin, sortmap[idx.upper_bound]);
+        cudf::interpolate::linear<TResult>(value, a, b, idx.fraction);
```

Comment:
I'm not sure where this is defined, but it looks like the output is being passed via a non-const ref. Prefer that outputs be returned.